quickwit/quickwit-doc-mapper/src/routing_expression/mod.rs

```diff
@@ -69,24 +70,47 @@ fn hash_json_val<H: Hasher>(json_val: &JsonValue, hasher: &mut H) {
     }
 }
 
+fn extract_value(data: &JsonValue, keys: &[Cow<str>]) -> Option<JsonValue> {
+    match keys {
+        [key, rest @ ..] => {
+            if let JsonValue::Object(obj) = data {
+                if let Some(value) = obj.get(key.as_ref()) {
+                    extract_value(value, rest)
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        }
+        [] => Some(data.clone()), // No more keys, return the current value
+    }
+}
+
 impl RoutingExprContext for serde_json::Map<String, JsonValue> {
     fn hash_attribute<H: Hasher>(&self, attr_name: &str, hasher: &mut H) {
-        if let Some(json_val) = self.get(attr_name) {
-            hasher.write_u8(1u8);
-            hash_json_val(json_val, hasher);
+        if let Ok(keys) = expression_dsl::parse_keys(attr_name) {
```

Comment:
`parse_keys` is still allocating a Vec for its return value (and we re-do parsing of that thing for each document, despite it never changing). I think you should change the prototype of  `RoutingExprContext::hash_attribute` to accept a slice of string (or whatever else is convenient), and store the corresponding Vec directly in `InnerRoutingExpr::Field`